### PR TITLE
Communities and groups

### PR DESCRIPTION
**Pull request by aapeliv**
This is a start, would be happy for comments.

I know the APIs are a bit inconsistent right now, and will require some refactoring. We'll need to write some kind of style guide for the APIs and clean them up eventually. I've been reading some Google's "API improvement proposals" which might be a good place to look for inspiration (e.g. https://google.aip.dev/158).

There's two models in a way, the [database/conceptual model](https://wiki.couchers.org/en/community/concept); but then presumably there will be some slightly different view of things that the users see (e.g. where points of interest and guides are in separate parts of the UI). How do we build the API? Presumably the backend should be the one to do this translation and deal with the difference.

### CODE REVIEW

File: app/pb/communities.proto
Comment (kalvdans) [inline app/pb/communities.proto]: I guess you mean `repeated ClusterMember` here.

**Review by kalvdans — COMMENTED** [review]
Sounds good to use https://google.aip.dev/158 as the pagination standard.

File: app/pb/communities.proto
Comment (aapeliv) [reply app/pb/communities.proto]: Yes, thanks

**Comment by lucaslcode** [comment]
The reasoning for pagination tokens sounds good, but it seems like so much overhead. How would we implement it? In a table?

**Comment by kalvdans** [comment]
> The reasoning for pagination tokens sounds good, but it seems like so much overhead. How would we implement it? In a table?

No, just the number of items to skip, as a decimal string. Or the ID to start at, as a decimal string.

**Comment by aapeliv** [comment]
> > The reasoning for pagination tokens sounds good, but it seems like so much overhead. How would we implement it? In a table?
> 
> No, just the number of items to skip, as a decimal string. Or the ID to start at, as a decimal string.

Yeah, I think the actual meaning of page tokens can be API-specific. But how they function should be standardised, they should have standard names, etc. I don't think we need to go through the trouble of encrypting/making them opaque though, since we don't have enough API consumers to have those issues.

**Comment by aapeliv** [comment]
I've discussed this with @ItsiW: the backend API will need to translate between the database structure and the user-facing structure. So the API will not contain terms such as "node" and "cluster", they'll have "communities", "groups", etc. I'll make some changes to this PR.

**Comment by aapeliv** [comment]
This is starting to be ready for review now. It's missing tests that I'll add once the API has been agreed on.

The code is a bit messy and will require refactoring at some point, but it works pretty well, and I want to wait till more of the communities features are built to see how code should be reorganized.

Currently RPCs that return paginated lists of users return IDs, and the other `List*` RPCs return full objects. @lucaslcode: should I use the former for everything? I don't know what frontend gymnastics are required for that, maybe that can be changed in another PR on `communities`? Also, I'm not sure if the `CommunityParent`, `GroupParent`, and `Parent` stuff is just messy and I should instead return IDs? It was just easy for me to get cute breadcrumbs like this...

File: app/pb/communities.proto
Comment (kalvdans) [inline app/pb/communities.proto:58]: Please describe more what is the slug.

**Comment by aapeliv** [comment]
It's very crude at the moment, but looks like this:

A community page:

Screen Shot 2021-01-11 at 7 33 00 pm
Screen Shot 2021-01-11 at 7 33 01 pm

A group page:

Screen Shot 2021-01-11 at 7 32 53 pm

File: app/pb/communities.proto
Comment (kalvdans) [inline app/pb/communities.proto:25]: ListCommunityAdmins to not collide with Group admins. Or is it namespaced everywhere?

Comment (aapeliv) [reply app/pb/communities.proto:25]: Yeah, group stuff is in `groups.proto` (in package `org.couchers.api.groups`, this is `org.couchers.api.communities`).

File: app/pb/communities.proto
Comment (aapeliv) [reply app/pb/communities.proto:58]: Done. I'd prefer to document it in a more central location as these will eventually be generated for basically every entity with a URL, but I've for now added the same comment to each `slug` field.

File: app/pb/communities.proto
Comment (aapeliv) [reply app/pb/communities.proto:25]: I think it's cleaner to not add `Community` to each of these RPCs and request/response messsages?

Comment (kalvdans) [reply app/pb/communities.proto:25]: Since everything is namespaced I think it is nicer with the same name yes. Resolving thread.

**Comment by aapeliv** [comment]
@kalvdans: do you approve the backend part?

File: app/backend/src/couchers/servicers/groups.py
Comment (kalvdans) [inline app/backend/src/couchers/servicers/groups.py]: Remove underscore or do not import internal functions.

File: app/pb/communities.proto
Comment (kalvdans) [reply app/pb/communities.proto:58]: Thanks. Since the slugify library is available in javascript as well, can't we leave it up to the frontends to generate the slug? Not a big deal though.

**Comment by kalvdans** [comment]
> @kalvdans: do you approve the backend part?

Yep so far :) Awaiting tests.

File: app/backend/src/couchers/servicers/groups.py
Comment (aapeliv) [reply app/backend/src/couchers/servicers/groups.py]: Fixed for now by removing the underscore/internalness. This is something that needs to be evaluated in the long run. Also awaiting Lucas's comments on whether this should just return page ids instead of full pages.

**Comment by lucaslcode** [comment]
I'm leaning towards yes, include the full parents.

We can manually insert them into the query cache to have instant loads when navigating to the parents. If we want to avoid bloating the cache, we can probably do something clever and convert that field to IDs before storing.

Actually to start with we don't even need to bother with any of that, not worth all that effort to avoid the load of a single api call.

File: app/frontend/src/features/communities/CommunityPage.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/CommunityPage.tsx:20]: These should probably `useSafeState` with `useIsMounted` as it's possible to navigate away during the async call, so they might try to set state on an unmounted component. Same for any state which is changed in an async function.

Comment (lucaslcode) [reply app/frontend/src/features/communities/CommunityPage.tsx:20]: Maybe you don't need to bother though, this will all be replaced with react-query hooks.

File: app/frontend/src/features/communities/CommunityPage.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/CommunityPage.tsx:266]: These can just be null.

**Review by lucaslcode — APPROVED** [review]
Looks good for getting it working, I don't think I need to nitpick on stuff we will be completely changing

**Comment by aapeliv** [comment]
I'll merge this in a bit, and tests can be written a bit later on `communities` before merge into `develop`.

**Pull request merged**
